### ISSUE TITLE
[FEATURE IDEA] Remote integrations ie decouple integrations registration from core code

### ISSUE THREAD

**Issue by r0psteev** [TracecatHQ/tracecat#287]
Hello, would be nice if we could build integrations that won't run on the same host
where core tracecat is deployed for a number of reasons (network isolation, scaling..).

Current way of developing integrations involves adding a few python scripts in the directory `./tracecat/actions/integrations/`
and registering methods they expose with `@registry.register` decorator.

For a scenario where the integration is running on a remote system, we could:

1. Have a dedicated python package for registering remote integrations, ex: `tracecat[remote_registry]`
2. Make this package implement registration over tracecat's api
3. Use it as follows

```python
# pip install tracecat[remote_registry]
from tracecat.remote_registry import registry

@registry.register(..., api_host="", api_port="")
def func():
     pass
```

**Comment by topher-lo** [comment]
Makes a ton of sense. Shouldn't to difficult to implement. Just need to update the docs to explain how this can be done. 

We also have a GitHub repo integration on the roadmap. Haven't had a chance to scope it in until after DEFCON mid August.

How the GitHub integration will work is you point the app to your repo with limited scope (read) and we will automatically sync integrations into your own tenant.

This requires teams to be built (expected release next week) #280

**Comment by r0psteev** [comment]
🫡

**Comment by mattdurant** [comment]
To add to this, something that would really elevate this feature is if the registration to a remote system is initiated by the remote side through a reverse tunnel, ala Azure App Proxy.  The use case I have in mind is you want to use Tracecat in the cloud, but you have a few on-prem systems you want to pull data from or interact with, so you can install the "remote" side on-prem and have it connect out to the cloud "main" instance and establishing a persistent SSL tunnel for receiving workflow requests.

**Comment by r0psteev** [comment]
Hello @mattdurant ,

Registration from remote systems via reverse tunnel is really good, as it will allow for registration at runtime,
and free users from having to track by themselves all these systems via a static config file for example.

But from what i understand; you think i am trying to shard workflow execution over multiple on-prem workers.

This isn't so much the concern, imho the current architecture for executing the workflow (i think  its based on temporal)
is pretty good.

The concern is at the level of the integrations blocks themselves.
Take terraform and it plugins for examples, the plugins are self-contained independent binaries that expose a unique
rpc interface, which allows the terraform core to use them transparently.

If integrations in tracecat (particularly custom ones) could be self contained (e.g as container deployments for example)
and expose a standard grpc interface, which tracecat calls at runtime when executing the workflows,
it will allow to have integrations be deployed on independent systems (subject to different network or ressource policies),
and even written in entirely different languages.

I don't know how feasible it is and how much complexity it could bring in the current code which is already very good.

Practically, the use case could be as follows:

- A developer builds a stateful integration and packages  it with its dependencies with a `docker-compose.yml`.
- This integration exposes a number of methods via grpc.
- Tracecat is some how given static config information (or registration via reverse tunnel as you proposed) about this integration and methods it exposes.
- During workflow execution, the methods the concerned integration exposes are called on its rpc.

I hope it isn't too obscure.

**Comment by mattdurant** [comment]
> it will allow to have integrations be deployed on independent systems (subject to different network or ressource policies), and even written in entirely different languages.

We're talking about similar benefits.  I was offering up an additional idea to expand the scope of this feature to allow what is a fairly standard offering by many SaaS products that need to interact with on-prem resources, which is a connector that would be installed in the restricted network and it phones home to the SaaS product to establish a communication channel.

I imagine that down the road there may be a hosted offering for Tracecat, and having a connectivity option to on-prem networks without needing to open inbound ports on your firewall will be a make-or-break feature for a lot of people.

**Comment by r0psteev** [comment]
Yes totally @mattdurant .
A connector that phones home to the cloud based SaaS instance would be the least intrusive
approach for Tracecat.
I just hope data movement /pipelining accross udfs (user defined functions) over this distributed system won't get
overly complicated.

**Comment by topher-lo** [comment]
> I imagine that down the road there may be a hosted offering for Tracecat, and having a connectivity option to on-prem networks without needing to open inbound ports on your firewall will be a make-or-break feature for a lot of people.

@lued is this related to adding Cloudflare tunnelling? Tines has this as an enterprise feature but I believe you mentioned it's not that hard to add (and use caddy to configure?). I'd like to open sauce something here. 

We could have a mini PoC after adding:
- External folder
- GitHub integration
- S3 bucket?

Options for storing your custom integrations.

@daryllimyt bump.

Prior art: https://docs.prefect.io/latest/guides/deployment/storage-guide/

**Comment by lued** [comment]
This is definitely endgame discussion to have this built-into the UI. @topher-lo, our discussion around cloudflare tunnels was during the reverse proxy decisions - so that is unrelated at the moment. 

I did see a Tines "self host" option which is likely similar to this topic. I did build a diagram for this integration in the future. Likely good to implement through a UI option in the future to initiate a "clustered" connection. 

If I understand the request correctly, I'd say we could build a POC through wireguard tunnels (or other open source STUN tunnel solutions) via docker containers and their overlay networks.

**Comment by topher-lo** [comment]
@lued might be unrelated but here's the tines feature I was referring to https://explained.tines.com/en/articles/6883952-deploying-and-using-the-tines-tunnel

**Comment by mattdurant** [comment]
IMO, if you are self-hosting, you've already figured out how to expose this to the Internet so you can use webhooks.  The functionality I was imagining was a way to remotely establish a persistent connection with an existing Tracecat instance and act as a remote action/workflow executor.

Something simple like a Websocket server on the Tracecat "server", and the "client" side makes an outbound connection via wss://, does authentication, and then messages can flow back and forth.

**Comment by topher-lo** [comment]
UPDATE: lots of transformational work being done here that'll shape the future of Tracecat. We're calling the decoupled integrations feature "Tracecat Registry".

TLDR: integrations registration is now a separate API and library (decoupled from the API service). We're also introducing integrations sync from private GitHub repos.

We are very open to scoping in reverse tunneling for integrations registration in self-hosted / airgapped environments. Will create a separate PR to continue this discussion.

- #403
- #404 
- #405 
- #408

**Comment by topher-lo** [comment]
Closed by registry (to be released in 0.11). As for the more sophisticated ideas around on-premise remote UDF deployments. Let's bring this into a new Issue!
